File: tests/omp/barrier_regions.cpp
void work(int n);

void sub3(int n)
{
  work(n);
  work(n);
}

void sub2(int k)
{
    sub3(k);
}

void sub1(int n)
{
  {
    for (int i=0; i<n; i++)
      sub2(i);
  }
}

int main()
{
  sub1(2);
  sub2(2);
  sub3(2);
  return 0;
}

File: tests/omp/barrier_regions.ml
open Optitrust
open Target

let _ = Run.script_cpp (fun _ ->
  !! Omp.barrier [tAfter; occIndex ~nb:2 0; cFunDef "sub3"; cCall "work"];
  !! Omp.parallel [Shared ["k"]] [tBefore; cFunDef "sub2";cCall "sub3"];
  !! Omp.parallel [Private ["i"]; Shared ["n"]] [tBefore; cSeq ~args:[[cFor "i"]] ()];
  !! Omp.for_ [] [tBefore; cFor "i"];
)
